### ISSUE TITLE
Generic tnt.Engine?

### ISSUE THREAD

**Issue by karandwivedi42** [pytorch/tnt#17]
Shouldn't the `tnt.engine.Engine` be changed to `tnt.engine.SGDEngine` and `tnt.engine.Engine` be a generic engine?

This is similar to [torchnet](https://github.com/torchnet/torchnet/tree/master/engine) which allowed extending engines to make meta-engines like `train_val_engine`.

**Comment by szagoruyko** [comment]
@karandwivedi42 Hi Karan! `tnt.engine.Engine` should be a generic engine, as it accepts any optimizer to be plugged in. I think we can add `train_val_engine` already.

**Comment by ndronen** [comment]
What is `train_val_engine`?  I ask because the `mnist.py` example for tnt doesn't cover using a validation set, and I'm curious how tnt is commonly used with a train-validation-test split.

**Comment by karandwivedi42** [comment]
@szagoruyko Hi! Is it okay if I send a PR for this?

**Comment by szagoruyko** [comment]
@karandwivedi42 sure, a PR is very much welcome!
@ndronen you can run validation in `on_end_epoch` hook, here's an example: https://github.com/szagoruyko/attention-transfer/blob/master/cifar.py#L322-L346

**Comment by szagoruyko** [comment]
I'll modify the mnist example to do that as well.
